File: 424. Longest Repeating Character Replacement/optimised.ts
function characterReplacement(s: string, k: number): number {
	const count = {};
	let res = 0;
	
	let l = 0;
	for (let r = 0; r < s.length; r++) {
		count[s[r]] = (count[s[r]] || 0) + 1;

		while ((r - l + 1) - Math.max(...Object.values(count)) > k) {
			count[s[l]] -= 1;
			l += 1;
		}

		res = Math.max(res, r - l + 1);
	}

	return res
};

console.log(characterReplacement("ABAB", 2));
console.log(characterReplacement("AABABBA", 1));

File: 424. Longest Repeating Character Replacement/optimised.py
#!/usr/bin/env python

def characterReplacement(s: str, k: int) -> int:
    n = len(s)
    count = {}
    res = 0
    
    l = 0
    maxf = 0
    for r in range(n):
        count[s[r]] = 1 + count.get(s[r], 0)
        maxf = max(maxf, count[s[r]])

        while (r - l + 1) - maxf > k:
            count[s[l]] -= 1
            l += 1

        res = max(res, r - l + 1)

    return res


print(characterReplacement("ABAB", 2))
print(characterReplacement("AABABBA", 1))
